include/nbl/builtin/hlsl/workgroup/bitonic_sort.hlsl

```diff
@@ -0,0 +1,253 @@
+        subgroup::bitonic_sort<SortConfig>::__call(subgroupAscending, lopair, hipair);
+
+        const uint32_t subgroupInvocationID = glsl::gl_SubgroupInvocationID();
+
+        [unroll]
+        for (uint32_t stage = 0; stage < numSubgroupsLog2; ++stage)
+        {
+            const bool bitonicAscending = !bool(invocationID & (subgroupSize << (stage + 1)));
+
+            mergeStage(sharedmemAccessor, stage, bitonicAscending, invocationID, lopair, hipair);
+
+            subgroup::bitonic_sort<SortConfig>::mergeStage(subgroupSizeLog2, bitonicAscending, subgroupInvocationID, lopair, hipair);
+        }
+
+        accessor.template set<nbl::hlsl::pair<key_t, value_t> >(loIdx, lopair);
+        accessor.template set<nbl::hlsl::pair<key_t, value_t> >(hiIdx, hipair);
+    }
+};
+
+// ==================== ElementsPerThreadLog2 > 1 Specialization (Virtual Threading) ====================
+// This handles larger arrays by combining global memory stages with recursive E=1 workgroup sorts
+template<uint16_t ElementsPerThreadLog2, uint16_t WorkgroupSizeLog2, typename KeyType, typename ValueType, typename Comparator, class device_capabilities>
+struct BitonicSort<bitonic_sort::bitonic_sort_config<ElementsPerThreadLog2, WorkgroupSizeLog2, KeyType, ValueType, Comparator>, device_capabilities>
+{
+    using config_t = bitonic_sort::bitonic_sort_config<ElementsPerThreadLog2, WorkgroupSizeLog2, KeyType, ValueType, Comparator>;
+    using simple_config_t = bitonic_sort::bitonic_sort_config<1, WorkgroupSizeLog2, KeyType, ValueType, Comparator>;
+
+    using key_t = KeyType;
+    using value_t = ValueType;
+    using comparator_t = Comparator;
+
+    template<typename Accessor, typename SharedMemoryAccessor>
+    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor)
+    {
+        const uint32_t WorkgroupSize = config_t::WorkgroupSize;
+        const uint32_t ElementsPerThread = config_t::ElementsPerInvocation;
+        const uint32_t TotalElements = WorkgroupSize * ElementsPerThread;
+        const uint32_t ElementsPerSimpleSort = WorkgroupSize * 2;
+
+        const uint32_t threadID = glsl::gl_LocalInvocationID().x;
+        comparator_t comp;
+
+        // PHASE 1: Sub-sorts in chunks of size WorkgroupSize*2
+        accessor_adaptors::Offset<Accessor> offsetAccessor;
+        offsetAccessor.accessor = accessor;
+
+        const uint32_t numSub = TotalElements / ElementsPerSimpleSort;
+
+        [unroll]
+        for (uint32_t sub = 0; sub < numSub; sub++)
+        {
+            if (sub)
+                sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
+
+                offsetAccessor.offset = sub * ElementsPerSimpleSort;
+
+                // Call E=1 workgroup sort
+                BitonicSort<simple_config_t, device_capabilities>::template __call(offsetAccessor, sharedmemAccessor);
+        }
+        sharedmemAccessor.workgroupExecutionAndMemoryBarrier();
```

Comment:
put the barrier at the end of the loop then you don't need `if(sub)` and the barrier after the loop